packages/ruleset/src/functions/use-date-based-format.js

```diff
@@ -0,0 +1,359 @@
+      if (schema.example) {
+        logger.debug(
+          `${ruleId}: adding example for schema at location: ${path.join('.')}`
+        );
+
+        examples.push({
+          example: schema.example,
+          examplePath: propertyPath.slice(), // Use a copy to prevent modification.
+        });
+      }
+
+      // Add sentinels for arrays/dictionaries to the running path,
+      // to assist the example-parsing logic later on. This must come
+      // after we push the example to the list.
+      if (isSchemaProperty(path)) {
+        if (isArraySchema(schema)) {
+          propertyPath.push('[]');
+        }
+
+        if (isDictionarySchema(schema)) {
+          propertyPath.push('{}');
+        }
+      }
+    }
+
+    // Use a slice (a copy) of the `propertyPath` array so that the
+    // invoked function can modify it without modifying the original.
+    return performValidation(
+      schema,
+      path,
+      apidef,
+      propertyPath.slice(),
+      examples
+    );
+  });
+}
+
+// This function performs the actual checks against a schema to determine if
+// it should be a "date" or "date-time" schema, but isn't defined as one.
+// It is wrapped in the outer function for the gathering of examples, etc. but
+// this function implements the checks: 1) see if the name of a property
+// indicates that it is a date-based schema and 2) see if the example value for
+// a schema indicates that it is a date-based schema.
+function performValidation(schema, path, apidef, propertyPath, examples) {
+  // If this is already a date or date-time schema, no need to check if it should be.
+  if (isDateSchema(schema) || isDateTimeSchema(schema)) {
+    logger.debug(
+      `${ruleId}: skipping date-based schema at location: ${path.join('.')}`
+    );
+
+    return [];
+  }
+
+  // Check for a name that would indicate the property should be date-based.
+  const hasDateTimeName =
+    isSchemaProperty(path) && isDateBasedName(path.at(-1));
+
+  logger.debug(
+    `${ruleId}: property at location: ${path.join('.')} has a date-based name`
+  );
+
+  if (hasDateTimeName && (isStringSchema(schema) || isIntegerSchema(schema))) {
+    // If the schema is determined to be a date-time schema by the name alone,
+    // we can return - no need to look for an example value.
+    return [
+      {
+        message:
+          'According to its name, this property should use type "string" and format "date" or "date-time"',
+        path,
+      },
+    ];
+  }
+
+  // Check example values for string schemas.
+  if (isStringSchema(schema)) {
+    // If this is a property, we need to include its name in the path.
+    if (isSchemaProperty(path)) {
+      propertyPath.push(path.at(-1));
+    }
+
+    // Either use the schema example directly or search the list of examples
+    // for an example object that contains a value for this property.
+    const exampleValue = schema.example || findExample(propertyPath, examples);
```

Comment:
This isn't composition-aware, so it will miss examples in schemas like this:
```
{
  "type": "string",
  "allOf": [
    {
      "example": "2024-12-19"
    }
  ]
}
```